### ISSUE TITLE
[Question or Bug] Extra Useless Tensor in grad_map

### ISSUE THREAD

**Issue by yewentao256** [darglein/TinyTorch#1]
Hello!

I have noticed that here the grad is always size of 2, but only grad[0] is used and has value.

```c++
struct AddNode : public FunctionNode<AddNode>
{
    static std::vector<Tensor> backward(Context& ctx, std::vector<Tensor> grad)
    {
        auto grad_a = add_backward_impl(grad[0]);
        return grad_a;
    }
};
```

This is because grad_map[next_node].resize(num_inputs) and num_inputs is 2(for add node)
```c++
          grad_map[next_node].resize(next_node->num_inputs);
          grad_map[next_node][next->input_nr].AddInplace(next_gradients[next->input_nr]);
```

But I try to print the `next->input_nr` and find it that it is always 0.

It seems the `input_nr` comes from here when building edge, but since forward result size is always 1, so input_nr is always 0
```c++
        // Forward
        auto result = T::forward(node->context, t);

        // Set the edges of the output to point to this node
        for (int i = 0; i < result.size(); ++i)
        {
            result[i].SetEdge(std::make_shared<Edge>(node, i));
        }
        return result;
```

Here comes the problem
1. grad_map's tensor list normally has two Tensor, but only the first one is used, the other one is size 0.
2. assert add's backward input grad list's size == 1 will fail, since it has two values. 

I am wondering if you could tell me is this a bug or not? If so, could you solve this problem?

**Comment by darglein** [comment]
Hey!

This was indeed a bug. The `num_input` variable was incorrectly set to `t.size()`, but should have been set to `result.size()`.

I have fixed the bug 6abd68f0df265527b524ab2f72c833e31c160974 and renamed the `num_inputs` variable to be more expressive.

Thank you for writing the issue.

**Comment by yewentao256** [comment]
Thanks for fixing the bug~
